src/SkeletalTrapezoidationGraph.cpp

```diff
@@ -0,0 +1,323 @@
+                edge_from_3->twin->to = quad_mid->from;
+                if (count > 50)
+                {
+                    std::cerr << edge_from_3->from->p << " - " << edge_from_3->to->p << '\n';
+                }
+                if (++count > 1000) 
+                {
+                    break;
+                }
+            }
+
+            // o-o > collapse top
+            // | |
+            // | |
+            // | |
+            // o o
+            if (quad_mid->from->some_edge == quad_mid)
+            {
+                if (quad_mid->twin->next)
+                {
+                    quad_mid->from->some_edge = quad_mid->twin->next;
+                }
+                else
+                {
+                    quad_mid->from->some_edge = quad_mid->prev->twin;
+                }
+            }
+            
+            nodes.erase(node_locator[quad_mid->to]);
+
+            quad_mid->prev->next = quad_mid->next;
+            quad_mid->next->prev = quad_mid->prev;
+            quad_mid->twin->next->prev = quad_mid->twin->prev;
+            quad_mid->twin->prev->next = quad_mid->twin->next;
+
+            safelyRemoveEdge(quad_mid->twin, edge_it, edge_it_is_updated);
+            safelyRemoveEdge(quad_mid, edge_it, edge_it_is_updated);
+        }
+
+        //  o-o
+        //  | | > collapse sides
+        //  o o
+        if ( should_collapse(quad_start->from, quad_end->to) && should_collapse(quad_start->to, quad_end->from))
+        { // Collapse start and end edges and remove whole cell
+
+            quad_start->twin->to = quad_end->to;
+            quad_end->to->some_edge = quad_end->twin;
+            if (quad_end->from->some_edge == quad_end)
+            {
+                if (quad_end->twin->next)
+                {
+                    quad_end->from->some_edge = quad_end->twin->next;
+                }
+                else
+                {
+                    quad_end->from->some_edge = quad_end->prev->twin;
+                }
+            }
+            nodes.erase(node_locator[quad_start->from]);
+
+            quad_start->twin->twin = quad_end->twin;
+            quad_end->twin->twin = quad_start->twin;
+            safelyRemoveEdge(quad_start, edge_it, edge_it_is_updated);
+            safelyRemoveEdge(quad_end, edge_it, edge_it_is_updated);
+        }
+        // If only one side had collapsable length then the cell on the other side of that edge has to collapse
+        // if we would collapse that one edge then that would change the quad_start and/or quad_end of neighboring cells
+        // this is to do with the constraint that !prev == !twin.next
+
+        if (!edge_it_is_updated)
+        {
+            edge_it++;
+        }
+    }
+}
+
+void SkeletalTrapezoidationGraph::makeRib(edge_t*& prev_edge, Point start_source_point, Point end_source_point, bool is_next_to_start_or_end)
+{
+    Point p = LinearAlg2D::getClosestOnLineSegment(prev_edge->to->p, start_source_point, end_source_point);
```

Comment:
If I understood correctly, we can use `getClosestOnLine` here, rather than `getClosestOnLineSegment`. We've already checked bounds of the projection so we don't need to do that again. See my previous comment here: https://github.com/Ultimaker/CuraEngine/pull/1210#discussion_r391073165

Reply:
In the sense that it wouldn't make much sense to have a rib that is not connected to the line segment (partially) defining it, and we have already committed to making the rib here, right?